### PR TITLE
Rearrange GPIO API

### PR DESCRIPTION
**Pull request by hs0225**
- rename Gpio API
  setPin -> open, writePin -> write, readPin -> read
- delete port API
- delete pin limit (related issues : https://github.com/Samsung/iotjs/issues/341)

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

File: src/platform/iotjs_module_gpio-linux-general.inl.h [hunk 1]
```diff
@@ -458,129 +406,54 @@ void SetPinWorker(uv_work_t* work_req) {
...
+  DDDLOG("GPIO WriteGpioWorker() - pin: %d, value: %d",
+         pin, value);
 
-void ReadPinWorker(uv_work_t* work_req) {
-  GpioLinuxGeneral* gpio = GpioLinuxGeneral::GetInstance();
-  IOTJS_ASSERT(gpio->_initialized == true);
-
-  GpioReqWrap* gpio_req = reinterpret_cast<GpioReqWrap*>(work_req->data);
-  GpioReqData* req_data = gpio_req->req();
+  char value_path[64] = {0};
```
Comment (nova0821) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: I would prefer to use ``#define`` for value ``64``.

File: src/platform/iotjs_module_gpio-linux-general.inl.h [hunk 2]
```diff
@@ -458,129 +406,54 @@ void SetPinWorker(uv_work_t* work_req) {
...
-
-  GpioReqWrap* gpio_req = reinterpret_cast<GpioReqWrap*>(work_req->data);
-  GpioReqData* req_data = gpio_req->req();
-
-  DDDLOG("GPIO ReadPortWorker() - port: %d" ,req_data->pin);
-
-  req_data->value = 0x00;
-  int32_t pin = req_data->pin * 8;
-  int32_t value = 0;
+  char value_path[64] = {0};
```
Comment (nova0821) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: ditto.

File: test/run_pass/test_gpio2.js [hunk 3]
```diff
@@ -18,55 +18,37 @@ var assert = require('assert');
...
+  gpio.open(LED_PIN_NUMBER, "out", LED_PIN_MODE);
+  gpio.open(SWITCH_PIN_NUMBER, "in", SWITCH_PIN_MODE);
 });
 
-gpio.on('setPort', function(port, dir, mode) {
-  console.log('setPort complete - port: %d, direction: %s, mode: %s',
-              port, dir, mode);
+gpio.on('open', function(pin, dir, mode) {
+  console.log('open complete - pin: %d, direction: %s, mode: %s',
+    pin, dir, mode);
```
Comment (nova0821) [inline test/run_pass/test_gpio2.js]: Please check indent.

File: docs/IoT.js-API-GPIO.md [hunk 4]
```diff
@@ -33,7 +42,7 @@ After this function any other GPIO function call except `initialize()` will be f
 'release' event will be emitted after GPIO device is released.
 
 
-#### gpio.setPin(pinNumber, direction[, mode][, callback])
+#### gpio.open(pinNumber, direction[, mode][, callback])
```
Comment (nova0821) [inline docs/IoT.js-API-GPIO.md]: I think ,in this section, we need to explain why ``open`` API is get △ in Platform Support section.

**Comment by chokobole** [comment]
Please make commit message start with verb. like Arrange GPIO API, this is looking good when git log --oneline

File: src/js/gpio.js [hunk 5]
```diff
@@ -316,56 +151,140 @@ GPIO.prototype.release = function(callback) {
...
+
+    // Emits error.
+    if (err) {
+      self.emit('error', err);
+    } else {
+      self.emit('write', pinNumber, value);
+    }
+  };
+
+  // write pin a value.
```
Comment (zherczeg) [inline src/js/gpio.js]: Comment: Set pin value.

File: src/module/iotjs_module_gpio.cpp [hunk 6]
```diff
@@ -46,80 +46,15 @@ Gpio* Gpio::GetInstance() {
...
-  const iotjs_jval_t* arg1 = JHANDLER_GET_ARG(1, function); \
-  GpioReqWrap* req_wrap = new GpioReqWrap(arg1); \
-  GpioReqData* req_data = req_wrap->req(); \
-  \
-  req_data->pin = JHANDLER_GET_ARG(0, number); \
-  req_data->op = kGpioOpRead ## readType; \
-  \
-  Gpio* gpio = Gpio::GetInstance(); \
-  gpio->Read ## readType(req_wrap);
+
```
Comment (zherczeg) [inline src/module/iotjs_module_gpio.cpp]: Is READ_GPIO an empty define?

File: src/platform/iotjs_module_gpio-linux-general.inl.h [hunk 7]
```diff
@@ -458,129 +406,54 @@ void SetPinWorker(uv_work_t* work_req) {
...
-  GpioLinuxGeneral* gpio = GpioLinuxGeneral::GetInstance();
-  IOTJS_ASSERT(gpio->_initialized == true);
-
-  GpioReqWrap* gpio_req = reinterpret_cast<GpioReqWrap*>(work_req->data);
-  GpioReqData* req_data = gpio_req->req();
+  char value_path[64] = {0};
+  snprintf(value_path, 63, GPIO_PIN_FORMAT_VALUE, pin);
 
-  DDDLOG("GPIO ReadPinWorker() - pin: %d" ,req_data->pin);
+  char buffer[2] = {0};
```
Comment (zherczeg) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: Is this ensures the second value of the buffer is 0?

File: src/platform/iotjs_module_gpio-linux-general.inl.h [hunk 8]
```diff
@@ -458,129 +406,54 @@ void SetPinWorker(uv_work_t* work_req) {
...
+  char value_path[64] = {0};
+  snprintf(value_path, 63, GPIO_PIN_FORMAT_VALUE, pin);
 
-  for (int offset = 0; offset < 8; offset++) {
-    if (!ReadPin(pin + offset, &value)) {
-      req_data->result = kGpioErrSys;
-      return;
-    }
-    if (value) req_data->value |= (1 << offset);
+  char buffer[10] = {0};
```
Comment (zherczeg) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: Why this buffer is initialized?

File: docs/IoT.js-API-GPIO.md [hunk 9]
```diff
@@ -124,30 +91,18 @@ Emitted after GPIO device is successfully initialized.
 
 Emitted after GPIO device is successfully released.
 
-#### `'setPin(pin: Number, direction: String, mode:String)'`
+#### `'open(pin: Number, direction: String, mode:String)'`
```
Comment (chokobole) [inline docs/IoT.js-API-GPIO.md]: Please explain parameters type or role following lines as above. The form is not unified.

File: docs/IoT.js-API-GPIO.md [hunk 10]
```diff
@@ -124,30 +91,18 @@ Emitted after GPIO device is successfully initialized.
...
 
 Emitted after GPIO pin is set.
 
-#### `'writePin(pin: Number, value: Boolean)'`
+#### `'write(pin: Number, value: Boolean)'`
```
Comment (chokobole) [inline docs/IoT.js-API-GPIO.md]: ditto

File: docs/IoT.js-API-GPIO.md [hunk 11]
```diff
@@ -124,30 +91,18 @@ Emitted after GPIO device is successfully initialized.
...
 
 Emitted after GPIO write pin is finished.
 
-#### `'readPin(pin: Number, value: Boolean)'`
+#### `'read(pin: Number, value: Boolean)'`
```
Comment (chokobole) [inline docs/IoT.js-API-GPIO.md]: ditto

File: src/js/gpio.js [hunk 12]
```diff
@@ -21,13 +21,6 @@ var assert  = require('assert');
...
 
-var GpioSettingType = {
-  kGpioPin: 'Pin',
-  kGpioPort: 'Port',
-  getLowLetter: function(type) {
-    return type === this.kGpioPin ? 'pin' : 'port';
-  }
-};
 
 function GpioError(code, operation, message) {
```
Comment (chokobole) [inline src/js/gpio.js]: Is it lead to infinite recursive? and where `errno` comes from? I think the code should be like followings.
```javascript
this.name = 'GpioError'
this.message = message;
return this;
```

File: src/js/gpio.js [hunk 13]
```diff
@@ -57,175 +50,17 @@ function CreateGpioError(operation, errno) {
...
   this._initialized = false;
-
-  this._gpioControlWorker = new GpioControlWorker(this);
 }
 
 util.inherits(GPIO, eventEmiter);
 
+
 // gpio.initialize(callback)
 //  callback: Function(err: GpioError | null)
```
Comment (chokobole) [inline src/js/gpio.js]: Please remove one additional blank.

File: src/js/gpio.js [hunk 14]
```diff
@@ -316,56 +151,140 @@ GPIO.prototype.release = function(callback) {
...
+  // After write handler.
+  var afterWrite = function(errno) {
+    var err = CreateGpioError('write', errno);
+
+    // If callback was given, calls back.
+    if (util.isFunction(callback)) {
+      callback(err);
+    }
+
+    // Emits error.
```
Comment (chokobole) [inline src/js/gpio.js]: Please type `Emit`

File: src/js/gpio.js [hunk 15]
```diff
@@ -316,56 +151,140 @@ GPIO.prototype.release = function(callback) {
...
+
+  if (pinNumber < 0) {
+    throw new Error('Invalid pinNumber');
+  }
+
+  // After read handler.
+  var afterRead = function(errno, value) {
+    var err = CreateGpioError('read', errno);
+
+    // calls back.
```
Comment (chokobole) [inline src/js/gpio.js]: Call back

File: src/js/gpio.js (hunk 14 above)
Comment (chokobole) [inline src/js/gpio.js]: Emit error

File: src/js/gpio.js [hunk 16]
```diff
@@ -316,56 +151,140 @@ GPIO.prototype.release = function(callback) {
...
+      self.emit('open', pinNumber, direction, mode);
+    }
+  };
 
-// gpio.writePort(portNumber, value[, callback])
-// event: 'writePort'
-GPIO.prototype.writePort = function(portNumber, value, callback) {
-  this._gpioControlWorker.writeGpio(GpioSettingType.kGpioPort,
-                                    gpio.writePort, arguments);
+  // open pin configuration.
```
Comment (chokobole) [inline src/js/gpio.js]: Open pin configuration

File: docs/IoT.js-API-GPIO.md (hunk 4 above)
Comment (nova0821) [inline docs/IoT.js-API-GPIO.md]: @hs0225 Ok. I see.

File: src/platform/iotjs_module_gpio-linux-general.inl.h (hunk 1 above)
Comment (hs0225) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: I fixed.

File: src/platform/iotjs_module_gpio-linux-general.inl.h (hunk 2 above)
Comment (hs0225) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: done.

File: test/run_pass/test_gpio2.js (hunk 3 above)
Comment (hs0225) [inline test/run_pass/test_gpio2.js]: done.

File: src/js/gpio.js (hunk 5 above)
Comment (hs0225) [inline src/js/gpio.js]: I fixed it.

File: src/module/iotjs_module_gpio.cpp (hunk 6 above)
Comment (hs0225) [inline src/module/iotjs_module_gpio.cpp]: I deleted it.

**Comment by hs0225** [comment]
@chokobole @zherczeg @nova0821 I updated this PR. Please review again.

File: src/platform/iotjs_module_gpio-linux-general.inl.h (hunk 8 above)
Comment (hs0225) [inline src/platform/iotjs_module_gpio-linux-general.inl.h]: I fixed it.

**Comment by zherczeg** [comment]
LGTM

**Comment by yichoi** [comment]
LGTM

**Pull request merged**